website/src/pages/tech-specs.md

```diff
@@ -0,0 +1,350 @@
+
+1. Pick an instant in the timeline (last successful commit or a specific commit version explicitly queried) and set that the commit time to compute the list of files to read from. 
+2. For the picked commit time, compute all the file slices that belongs to that specific commit time. For all the partition paths involved in the query, the file slices that belong to an successful commit before the picked commit should be included. The lookup on the filesystem could be slow and inefficient and can be further optimized by caching in memory or using the files (partition path to filename) index or with a support of a external timeline serving system.  
+3. For the merge on read table type, ensure the apprpriate merging rules are applied to apply the updates queued for the base in the log files.
+   1. [TODO] List merging rules - (ordering field).
+
+
+
+## Writer Expectations
+
+Writer into Hudi will have to ingest new records, updates to existing records or delete records into the dataset. All transactional actions follow the same state transition as described in the transaction log (timeline) section. Writers will optimistically create new base ang log files and will finally transition the action state to completed to register all the modifications to the dataset atomically. Writer merges the data using the following steps
+
+1. Writer will pick a monotonically increasing instant time from the latest state of the Hudi timeline (**action commit time**) and will pick the last successful commit instant (**merge commit time**) to merge the changes to. If the merge succeeds, then action commit time will be the next successful commit in the timeline. 
+2. For all the incoming records, the writer will have to efficiently determine if this is an update or insert. This is done by a process called tagging - which is a batched point lookups of the row key and partition path pairs in the entire dataset. The efficiency of tagging is critical to the merge performance. This can be optimized with indexes (bloom, global key value based index) and caching. New records will not have a tag. 
+3. Once records are tagged, the writer can apply them onto the specific file slice. 
+   1. For copy on write, writer will create a new slice (action commit time) of the base file in the file group
+   2. For merge on read, writer will create a new log file with the action commit time on the merge commit time file slice
+4. Deletes are encoded as special form of updates where only the meta fields and the operation is populated. See the delete block type in log format block types. 
+5. Once all the writes into the file system is complete, concurrency control checks happen to ensure there are no overlapping writes and if that succeeds, the commit action is completed in the timeline atomically making the changes merged visible for the next reader.  
+
+
+
+## Table maintenance
+
+[TODO.- flush out what needs to be done for each of these]
```

Comment:
We can pick certain sections from Hudi docs, e.g. for clustering https://hudi.apache.org/docs/clustering#running-clustering
Let's discuss how much details we want here.